### PR TITLE
Dropout Vectorized Kernel 

### PR DESCRIPTION
**Pull request by ytaous**
**Description**: CUDA summary analysis for ULR-XL model shows room to improve performance in areas of dropout, biasdropout, and dropoutgrad CUDA kernels. 

image

With some investigation, we experiment it with vectorized data load/store approach in CUDA kernel, and the results shows ~0.75% of perf gain comparing to the original non-vectorized kernel logic in a short test (3k steps w/ full recipe on ITP cluster).

image

Another test with nvprof profiler also show consistent observation on a local run for 10 steps:

image

image

### CODE REVIEW

**Comment by ytaous** [comment]
will also run Python Packaging Pipeline to check actual loss with this PR

File: onnxruntime/contrib_ops/cuda/math/bias_dropout_impl.cu [hunk 1]
```diff
@@ -49,34 +49,82 @@ __global__ void BiasDropoutKernel(
...
-    float4 rand = curand_uniform4(&state);
-
-  #pragma unroll
-    for (CUDA_LONG i = 0; i < UNROLL; i++) {
-      CUDA_LONG li = id + gridDim.x * blockDim.x * i;
-      if (li < N) {
-        int offset = fdm_dim.mod(li);
+  float4 rand;
+
+  if (N % UNROLL != 0) {
```
Comment (SherlockNoMad) [inline onnxruntime/contrib_ops/cuda/math/bias_dropout_impl.cu:54]: I would suggest writing a separate Kernel for the vectorized impl. 
The selection of the kernel can be done at kernel launching time.

Comment (SherlockNoMad) [reply onnxruntime/contrib_ops/cuda/math/bias_dropout_impl.cu:54]: Same comments applies to Dropout and DropoutGrad kernels.

File: onnxruntime/core/providers/cuda/nn/dropout_impl.cu [hunk 2]
```diff
@@ -44,26 +44,60 @@ __global__ void DropoutKernel(
...
+    // From CUDA curand documentation:
+    //   The Philox_4x32_10 algorithm is closely tied to the thread and block count.
+    //   Each thread computes 4 random numbers in the same time thus the most efficient
+    //   use of Philox_4x32_10 is to generate a multiple of 4 times number of threads.
+    for (CUDA_LONG id = idx; id < rounded_size; id += step_size) {
+      rand = curand_uniform4(&state);
+    
+      #pragma unroll
+      for (CUDA_LONG i = 0; i < UNROLL; i++) {
+        CUDA_LONG li = id + gridDim.x * blockDim.x * i;
```
Comment (SherlockNoMad) [inline onnxruntime/core/providers/cuda/nn/dropout_impl.cu:61]: as Weixing suggested, vectorized and non-vectorized kernels are using different thread-to-id mapping... 
vectorized kernel are handling 4 contiguous elements per thread, non-vectorized kernels are handling 4 elements "step_size" apart... 
We should simplify this...

Comment (SherlockNoMad) [reply onnxruntime/core/providers/cuda/nn/dropout_impl.cu:61]: Same comments applied to Dropout, DropoutGrad and BiasDropout kernels.

Comment (ytaous) [reply onnxruntime/core/providers/cuda/nn/dropout_impl.cu:61]: If no perf improvement, we can leave the old access pattern as is. 
@weixingzhang

**Review by weixingzhang — APPROVED** [review]

File: onnxruntime/core/providers/cuda/nn/dropout_impl.cu [hunk 3]
```diff
@@ -44,26 +44,60 @@ __global__ void DropoutKernel(
...
+      #pragma unroll
+      for (int ii = 0; ii < UNROLL; ii++) {
+        mask[ii] = (&rand.x)[ii] < p;
+        r[ii] = T(float(src[ii]) * mask[ii] * scale);
+      }
+      // Vectorized writes for mask_data & Y_data
+      *(reinterpret_cast<LoadT*>(&Y_data[id])) = *reinterpret_cast<LoadT*>(&r[0]);
+      *(reinterpret_cast<MaskLoadT*>(&mask_data[id])) = *reinterpret_cast<MaskLoadT*>(&mask[0]);
+
+      __syncthreads();
```
[the remark this answers is not in the document]
Comment (SherlockNoMad) [reply onnxruntime/core/providers/cuda/nn/dropout_impl.cu]: hmm... this is still needed in the PT dropout impl... 
I think they are still needed ?

**Review by weixingzhang — APPROVED** [review]

**Review by baijumeswani — APPROVED** [review]

**Pull request merged**